java/vortex-jni/src/main/java/dev/vortex/impl/NativeDType.java

```diff
@@ -0,0 +1,170 @@
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package dev.vortex.impl;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.collect.ImmutableList;
+import com.sun.jna.Memory;
+import com.sun.jna.ptr.IntByReference;
+import dev.vortex.api.DType;
+import dev.vortex.jni.FFI;
+import java.nio.charset.StandardCharsets;
+import java.util.Optional;
+
+public final class NativeDType extends BaseWrapped<FFI.FFIDType> implements DType {
+
+    // Assumes no field name is > 1KiB
+    private static final int MAX_FIELD_LEN = 1024;
+    private static final int MAX_TIMEZONE_LEN = 2048;
+
+    private final Variant variant;
+
+    public NativeDType(FFI.FFIDType inner) {
+        super(inner);
+        this.variant = Variant.from(FFI.DType_get(inner));
+    }
+
+    @Override
+    public Variant getVariant() {
+        return variant;
+    }
+
+    /**
+     * True if the DType is nullable
+     */
+    @Override
+    public boolean isNullable() {
+        checkNotNull(inner);
+
+        return FFI.DType_nullable(inner);
+    }
+
+    @Override
+    public String toString() {
+        checkNotNull(inner);
+
+        var result = new StringBuilder();
+        result.append("DType{");
+        result.append(this.variant.toString());
+        result.append(", nullable=");
+        result.append(isNullable());
+
+        if (variant == Variant.STRUCT) {
+            result.append(", fields=[");
+            result.append(String.join(", ", getFieldNames()));
+            result.append("]");
+        }
+
+        // TODO: List, Extension type handling.
+
+        result.append("}");
+
+        return result.toString();
+    }
+
+    @Override
+    public ImmutableList<String> getFieldNames() {
+        checkNotNull(inner);
+        checkArgument(Variant.STRUCT == variant, "getStructFieldNames() for non-struct DType");
+
+        ImmutableList.Builder<String> builder = ImmutableList.builder();
+
+        // We assume no field name is >= 1KiB.
+        try (Memory memory = new Memory(MAX_FIELD_LEN)) {
+            var fieldCount = FFI.DType_field_count(inner);
+            for (int i = 0; i < fieldCount; i++) {
+                var lenRef = new IntByReference();
+                FFI.DType_field_name(inner, i, memory, lenRef);
+                var data = memory.getByteArray(0, lenRef.getValue());
+                var name = new String(data, StandardCharsets.UTF_8);
+                builder.add(name);
+            }
+        }
+
+        return builder.build();
+    }
+
+    @Override
+    public ImmutableList<DType> getFieldTypes() {
```

Comment:
The return type should be just the interace, for constants the type would matter here